// @easy-webpack/config-typescript/dist/index.js
"use strict";
var core_1 = require("@easy-webpack/core");
var path = require("path");
var _a = require('awesome-typescript-loader'), TsConfigPathsPlugin = _a.TsConfigPathsPlugin, CheckerPlugin = _a.CheckerPlugin; // does not complain about typings
module.exports = function typescript(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.options, options = _c === void 0 ? undefined : _c, _d = _b.exclude, exclude = _d === void 0 ? null : _d;
    return function typescript() {
        var loader = {
            test: /\.tsx?$/,
            loader: 'awesome-typescript-loader',
            exclude: exclude || (this.metadata.root ? [path.join(this.metadata.root, 'node_modules')] : []),
            options: options
        };
        return {
            resolve: {
                extensions: core_1.get(this, 'resolve.extensions', ['.js']).concat(['.tsx', '.ts'])
            },
            module: {
                rules: core_1.get(this, 'module.rules', []).concat([loader])
            },
            plugins: [
                new TsConfigPathsPlugin(options),
                new CheckerPlugin()
            ].concat(core_1.get(this, 'plugins', []))
        };
    };
};
